Dynamic Content Scaling is needlessly enabled for some canvas elements

hortont424 opened the pull request:
#### 89b7e1f4bfb2a61da2b79fc4d06de621187d1e67
<pre>
Dynamic Content Scaling is needlessly enabled for some canvas elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=287372">https://bugs.webkit.org/show_bug.cgi?id=287372</a>
<a href="https://rdar.apple.com/144489486">rdar://144489486</a>

Reviewed by Wenson Hsieh.

Dynamic Content Scaling doesn&apos;t work for canvas right now, because we don&apos;t
record and replay canvas commands. Because of this, we disable it for layers
that back canvases, to avoid pointless overhead.

However, to identify canvas layers, we were using isBitmapOnly(), which does not
include e.g. canvases that have backgrounds.

Instead, centralize the Dynamic Content Scaling policy in its own method, which
applies to all canvases regardless of decorations, *and* bitmap-only layers.

Also, the way that Dynamic Content Scaling enablement was plumbed to GPUP-hosted
layer backing was very duplicative (instead of conferring with includeDisplayList,
we would check isBitmapOnly() again!). Repair this so we can have the logic
in only one place. This also obviates the need for RenderingPurpose::BitmapOnlyLayerBacking,
which was a bit of a conflation of two separate dimensions.

* Source/WebCore/platform/graphics/RenderingMode.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::surfaceNameToNSString):
(WebCore::IOSurface::nameForRenderingPurpose):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::shouldUseRemoteRenderingFor):
Get rid of BitmapOnlyLayerBacking.

* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::layerAllowsDynamicContentScaling const):
(WebCore::GraphicsLayerClient::layerContainsBitmapOnly const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerAllowsDynamicContentScaling const):
(WebCore::PlatformCALayerClient::platformCALayerContainsBitmapOnly const): Deleted.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::layerAllowsDynamicContentScaling const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::allowsDynamicContentScaling const):
(WebKit::PlatformCALayerRemote::shouldIncludeDisplayListInBackingStore const):
(WebKit::PlatformCALayerRemote::containsBitmapOnly const): Deleted.
Replace &quot;contains bitmap only&quot; layer-level plumbing with &quot;allows dynamic content scaling&quot; one.
Let RenderLayer maintain the single implementation of the policy.

* Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h:
Hoist IncludeDynamicContentScalingDisplayList down here to fix some circular header
dependencies in WebKit.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::updateConfiguration):
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
(WebKit::RemoteImageBufferSet::isOpaque const):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):
* Source/WebKit/Shared/graphics/RemoteImageBufferSetConfiguration.h: Added.
* Source/WebKit/Shared/graphics/RemoteImageBufferSetConfiguration.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::setConfiguration):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
Encapsulate RemoteImageBufferSet configuration in a struct, so we can
easily have an optional bit for whether Dynamic Content Scaling is enabled or not.

Canonical link: <a href="https://commits.webkit.org/290181@main">https://commits.webkit.org/290181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50e9e510e6ea2047bdde7712c0a102a6a6a1de1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8703 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94160 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39940 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9091 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16892 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68719 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26388 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6972 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80943 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49081 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35321 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39047 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77084 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36309 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95993 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16359 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77595 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76729 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18962 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21296 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9474 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16373 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21684 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16114 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19565 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17895 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -239,14 +239,9 @@ void RemoteImageBufferSetProxy::close()
 }
 
-void RemoteImageBufferSetProxy::setConfiguration(WebCore::FloatSize size, float scale, const WebCore::DestinationColorSpace& colorSpace, WebCore::ImageBufferPixelFormat pixelFormat, WebCore::RenderingMode renderingMode, WebCore::RenderingPurpose renderingPurpose)
+void RemoteImageBufferSetProxy::setConfiguration(const RemoteImageBufferSetConfiguration& configuration)
```
```suggestion
void RemoteImageBufferSetProxy::setConfiguration(RemoteImageBufferSetConfiguration&& configuration)
```

whsieh commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -239,14 +239,9 @@ void RemoteImageBufferSetProxy::close()
-    m_pixelFormat = pixelFormat;
-    m_renderingMode = renderingMode;
-    m_renderingPurpose = renderingPurpose;
+    m_configuration = configuration;
```
```suggestion
    m_configuration = WTFMove(configuration);
```

whsieh commented on Source/WebKit/Shared/graphics/RemoteImageBufferSetConfiguration.h:
```diff
@@ -0,0 +1,56 @@
+    WebCore::FloatSize logicalSize;
+    float resolutionScale { 1.0f };
+    WebCore::DestinationColorSpace colorSpace { WebCore::DestinationColorSpace::SRGB() };
+    WebCore::ImageBufferPixelFormat pixelFormat;
```
Nit - perhaps we should use a default value here, to harden against UB?

hortont424 replied:
hmmmmm. I will default to BGRA8. I moved these defaults from the members, a bit surprising there was no default there!

whsieh commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
```diff
@@ -112,9 +112,18 @@
+            .includeDisplayList = m_parameters.includeDisplayList,
+#endif
+        };
+        m_bufferSet->setConfiguration(configuration);
```
```suggestion
        m_bufferSet->setConfiguration(WTFMove(configuration));
```

whsieh commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
```diff
@@ -100,7 +101,7 @@ class RemoteImageBufferSetProxy : public IPC::WorkQueueMessageReceiver, public I
     std::unique_ptr<ThreadSafeImageBufferSetFlusher> flushFrontBufferAsync(ThreadSafeImageBufferSetFlusher::FlushType);
 
-    void setConfiguration(WebCore::FloatSize, float, const WebCore::DestinationColorSpace&, WebCore::ImageBufferPixelFormat, WebCore::RenderingMode, WebCore::RenderingPurpose);
+    void setConfiguration(const RemoteImageBufferSetConfiguration&);
```
```suggestion
    void setConfiguration(RemoteImageBufferSetConfiguration&&);
```

mattwoodrow commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3228,6 +3228,12 @@ bool RenderLayerBacking::isBitmapOnly() const
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+bool RenderLayerBacking::layerAllowsDynamicContentScaling(const GraphicsLayer*) const
+{
+    return m_owningLayer.allowsDynamicContentScaling();
```
For WebGL/GPU, I think we usually end up with a 'contents' platform layer that holds the output frames directly, and then all the normal HTML content (borders, backgrounds etc) will be in a different platform layer.

I think this change will disable the dynamic content scaling for the background unnecessarily, and we could potentially fix that. 

Unsure if it matters though, since we also lose dynamic content scaling when the canvas contents get drawn into the same layer as the background (canvas 2d).

hortont424 replied:
I think "does not matter", since their contents will still not scale, just like 2D Canvas.

I did briefly consider doing something similar for 2D Canvas (contents in a different layer than decorations, to split them, so we could keep decorations scaling), but I don't think it's worth it.

hortont424 committed:
Dynamic Content Scaling is needlessly enabled for some canvas elements
https://bugs.webkit.org/show_bug.cgi?id=287372
rdar://144489486

Reviewed by Wenson Hsieh.

Dynamic Content Scaling doesn't work for canvas right now, because we don't
record and replay canvas commands. Because of this, we disable it for layers
that back canvases, to avoid pointless overhead.

However, to identify canvas layers, we were using isBitmapOnly(), which does not
include e.g. canvases that have backgrounds.

Instead, centralize the Dynamic Content Scaling policy in its own method, which
applies to all canvases regardless of decorations, *and* bitmap-only layers.

Also, the way that Dynamic Content Scaling enablement was plumbed to GPUP-hosted
layer backing was very duplicative (instead of conferring with includeDisplayList,
we would check isBitmapOnly() again!). Repair this so we can have the logic
in only one place. This also obviates the need for RenderingPurpose::BitmapOnlyLayerBacking,
which was a bit of a conflation of two separate dimensions.

* Source/WebCore/platform/graphics/RenderingMode.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::surfaceNameToNSString):
(WebCore::IOSurface::nameForRenderingPurpose):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::shouldUseRemoteRenderingFor):
Get rid of BitmapOnlyLayerBacking.

* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::layerAllowsDynamicContentScaling const):
(WebCore::GraphicsLayerClient::layerContainsBitmapOnly const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerAllowsDynamicContentScaling const):
(WebCore::PlatformCALayerClient::platformCALayerContainsBitmapOnly const): Deleted.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::layerAllowsDynamicContentScaling const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::allowsDynamicContentScaling const):
(WebKit::PlatformCALayerRemote::shouldIncludeDisplayListInBackingStore const):
(WebKit::PlatformCALayerRemote::containsBitmapOnly const): Deleted.
Replace "contains bitmap only" layer-level plumbing with "allows dynamic content scaling" one.
Let RenderLayer maintain the single implementation of the policy.

* Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h:
Hoist IncludeDynamicContentScalingDisplayList down here to fix some circular header
dependencies in WebKit.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::updateConfiguration):
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
(WebKit::RemoteImageBufferSet::isOpaque const):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):
* Source/WebKit/Shared/graphics/RemoteImageBufferSetConfiguration.h: Added.
* Source/WebKit/Shared/graphics/RemoteImageBufferSetConfiguration.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::setConfiguration):
(WebKit::RemoteImageBufferSetProxy::willPrepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
Encapsulate RemoteImageBufferSet configuration in a struct, so we can
easily have an optional bit for whether Dynamic Content Scaling is enabled or not.

Canonical link: https://commits.webkit.org/290181@main
